core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java

```diff
@@ -0,0 +1,343 @@
+
+        if (containsAnyStrategy(strategies)) {
+            configuration.strategies = strategies.split("\\s*,\\s*");
+        }
+
+        configuration.disable = Boolean.valueOf(overWriteSystemProperty(yamlConfiguration, "disable", SMART_TESTING_DISABLE, "false"));
+        configuration.debug = Boolean.valueOf(overWriteSystemProperty(yamlConfiguration, "debug", SMART_TESTING_DEBUG, "false"));
+        configuration.applyTo = overWriteSystemProperty(yamlConfiguration, "applyTo", SMART_TESTING_APPLY_TO, null);
+
+        final Map<String, Object> reportConfig = (Map<String, Object>) yamlConfiguration.get("report");
+
+        final Report report = new Report();
+        report.setEnable(Boolean.valueOf(overWriteSystemProperty(reportConfig, "enable", SMART_TESTING_REPORT_ENABLE, "false")));
+        report.setDir("target");
+        report.setName(REPORT_FILE_NAME);
+
+        configuration.report = report;
+
+        final Map<String, Object> scmConfig = (Map<String, Object>) yamlConfiguration.get("scm");
+
+        final String lastChanges = overWriteSystemProperty(scmConfig, "lastChanges", SCM_LAST_CHANGES, DEFAULT_LAST_COMMITS);
+
+        final Map<String, Object> scmRange = scmConfig != null ? (Map<String, Object>) scmConfig.get("range") : null;
+        final Range range = Range.builder()
+                .head(overWriteSystemProperty(scmRange, "head", SCM_RANGE_HEAD, HEAD))
+                .tail(overWriteSystemProperty(scmRange, "tail", SCM_RANGE_TAIL, String.join("~", HEAD, lastChanges)))
+            .build();
+
+        configuration.scm = Scm.builder().range(range).build();
+
+        return configuration;
+    }
+
+    private static String overWriteSystemProperty(Map<String, Object> yamlConfig, String key, String propertyName, String defaultValue) {
+        if (System.getProperty(propertyName) != null) {
+            return System.getProperty(propertyName);
+        } else if (yamlConfig != null && yamlConfig.get(key) != null) {
+            final Object value = yamlConfig.get(key);
+            if (value instanceof List) {
+                final List<String> list = (List<String>) value;
+                return String.join(",", list);
+            }
+            return String.valueOf(value);
+        } else {
+            return defaultValue;
+        }
+    }
+
+    private static boolean containsAnyStrategy(String strategies) {
+        return !strategies.trim().isEmpty();
+    }
+
+    public boolean isSelectingMode() {
+        return isModeSet() && RunMode.SELECTING == getMode();
+    }
+
+    public boolean isModeSet() {
+        return this.mode != null;
+    }
+
+    public boolean areStrategiesDefined() {
+        return strategies.length > 0;
+    }
+
+    public boolean isApplyToDefined() {
+        return this.applyTo != null;
+    }
+
+    public static Builder builder() {
+        return new Configuration.Builder();
+    }
+
+    public static class Builder {
```

Comment:
If I understand the purpose of the builders correctly, they are used only for test purposes. If so, then I would move them to the `TestBed` implementation to keep the class clear and to not expose test logic from the normal class.

Reply:
@MatousJobanek If we move it to TestBed then we can't use it in unit test for [ConfigurationTest](https://github.com/arquillian/smart-testing/pull/183/files#diff-09137b69519fe19b2f7c61862edeba9b) & [ConfigurationUsingPropertyTest](https://github.com/arquillian/smart-testing/pull/183/files#diff-ebaf055f2f7537b003d0b8833f51d445). Also as we are keeping it here it's quite easy to reference things in future if we want to modify it.

Reply:
I'm still not a fan of this. If it is a test logic, then I would move it to the test-bed module or make it as hidden as possible (package protected). 
If you want to open the API to the world, then you should also verify that the builder works properly - create a test for the builder. 
If you moved it to the test-bed, then in Core you could use different approach of creating the object:
* reading a file
* using the set methods (that you have there anyway)
* configuring the object by providing `Properties` object - that would make more sense exposing it than the static builder API